Virtual thread reference documentation and resteasy reactive how-to (#34913)
quarkusio/quarkus: docs/src/main/asciidoc/resteasy-reactive-virtual-threads.adoc

== Create the Maven project

First, we need a new project. Create a new project with the following command:

:create-app-artifact-id: rest-virtual-threads
:create-app-extensions: resteasy-reactive-jackson,quarkus-rest-client-reactive-jackson
include::{includes}/devtools/create-app.adoc[]

This command generates a new project importing the RESTEasy Reactive, Reactive REST client, and https://github.com/FasterXML/jackson[Jackson] extensions,
and in particular, adds the following dependencies:

[source,xml,role="primary asciidoc-tabs-target-sync-cli asciidoc-tabs-target-sync-maven"]
.pom.xml
----
<dependency>
  <groupId>io.quarkus</groupId>
  <artifactId>quarkus-resteasy-reactive-jackson</artifactId>
</dependency>
<dependency>
  <groupId>io.quarkus</groupId>
  <artifactId>quarkus-rest-client-reactive-jackson</artifactId>
</dependency>
----

[source,gradle,role="secondary asciidoc-tabs-target-sync-gradle"]
.build.gradle
----
implementation("io.quarkus:quarkus-resteasy-reactive-jackson")
implementation("quarkus-rest-client-reactive-jackson")
----

[NOTE]
====

> You may wonder why we use _reactive_ extensions.

Reviewer: ```suggestion
You might wonder why we use _reactive_ extensions.
```

Virtual threads have limitations, and we can only integrate them properly when using the reactive extensions.

> However, no worries, your code will be 100% synchronous.

Reviewer: ```suggestion
No worries: your code will be written 100% in a synchronous / imperative style.
```

Check the xref:./virtual-threads.adoc#why-not[virtual thread reference guide] for details.
====

== Prepare the `pom.xml` file

> We need to customize the `pom.xml` file to allow using virtual threads.

Reviewer: ```suggestion
We need to customize the `pom.xml` file to use virtual threads.
```

[source, xml]
----
<plugin>
  <artifactId>maven-surefire-plugin</artifactId>
  <version>${surefire-plugin.version}</version>
  <configuration>
    <systemPropertyVariables>
      <java.util.logging.manager>org.jboss.logmanager.LogManager</java.util.logging.manager>
      <maven.home>${maven.home}</maven.home>
    </systemPropertyVariables>
    <argLine>--enable-preview -Djdk.tracePinnedThreads</argLine> <!-- Added line -->
  </configuration>
</plugin>
<plugin>
  <artifactId>maven-failsafe-plugin</artifactId>
  <version>${surefire-plugin.version}</version>
  <executions>
    <execution>
      <goals>
        <goal>integration-test</goal>
        <goal>verify</goal>
      </goals>
      <configuration>
        <systemPropertyVariables>
          <native.image.path>${project.build.directory}/${project.build.finalName}-runner</native.image.path>
          <java.util.logging.manager>org.jboss.logmanager.LogManager</java.util.logging.manager>
          <maven.home>${maven.home}</maven.home>
        </systemPropertyVariables>
        <argLine>--enable-preview -Djdk.tracePinnedThreads</argLine> <!-- Added line -->
      </configuration>
    </execution>
  </executions>
</plugin>
----

> Note that the `--enable-preview` flag is only required if you use Java 19 or 20.

Reviewer: I'd extract this with an `important` admonition and make sure this pops out as it's quite critical.

> The `-Djdk.tracePinnedThreads` will detect pinned carrier threads while running tests (See xref:./virtual-threads#pinning[the virtual thread reference guide for details]).

Reviewer 2: +100 for this

== Create the weather client

This section is not about virtual threads.

> But, because we need to do some I/O to demonstrate virtual threads, we need a client doing I/O.

Reviewer: ```suggestion
Because we need to do some I/O to demonstrate virtual threads usage, we need a client doing I/O operations.
```

> In addition, the reactive REST client is virtual thread friendly (it does not pin and handle propagation correctly).

Reviewer: ```suggestion
In addition, the reactive REST client is virtual thread friendly: it does not pin and handle propagation correctly.
```

Create the `src/main/java/org/acme/WeatherService.java` class with the following content:

[source, java]
----
package org.acme;

import com.fasterxml.jackson.annotation.JsonProperty;
import io.quarkus.rest.client.reactive.ClientQueryParam;
import jakarta.ws.rs.GET;
import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;
import org.jboss.resteasy.reactive.RestQuery;

@RegisterRestClient(baseUri = "https://api.open-meteo.com/v1/forecast")
public interface WeatherService {

    @GET
    @ClientQueryParam(name = "current_weather", value = "true")
    WeatherResponse getWeather(@RestQuery double latitude, @RestQuery double longitude);

    record WeatherResponse(@JsonProperty("current_weather") Weather weather) {
        // represents the response
    }

    record Weather(double temperature, double windspeed) {
        // represents the inner object
    }
}
----

This class models the HTTP interaction with the weather service.

> Read more about the rest client on the dedicated xref:./rest-client-reactive[guide].

Reviewer: ```suggestion
Read more about the rest client in the dedicated xref:./rest-client-reactive[guide].
```

import io.smallrye.common.annotation.RunOnVirtualThread;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import org.eclipse.microprofile.rest.client.inject.RestClient;

@Path("/")
public class TheBestPlaceToBeResource {

    static final double VALENCE_LATITUDE = 44.9;
    static final double VALENCE_LONGITUDE = 4.9;

    static final double ATHENS_LATITUDE = 37.9;
    static final double ATHENS_LONGITUDE = 23.7;

    @RestClient WeatherService service;

    @GET
    @RunOnVirtualThread <1>
    public String getTheBestPlaceToBe() {
        var valence = service.getWeather(VALENCE_LATITUDE, VALENCE_LONGITUDE).weather().temperature();
        var athens = service.getWeather(ATHENS_LATITUDE, ATHENS_LONGITUDE).weather().temperature();

        // Advanced decision tree
        if (valence > athens && valence <= 35) {
            return "Valence! (" + Thread.currentThread() + ")";
        } else if (athens > 35) {
            return "Valence! (" + Thread.currentThread() + ")";
        } else {
            return "Athens (" + Thread.currentThread() + ")";
        }
    }
}
----
<1> Instructs Quarkus to invoke this method on a virtual thread

== Run the application in dev mode

> Make sure you use JDK and JVM supporting virtual thread and launch the _dev mode_ with `./mvnw quarkus:dev`:

Reviewer: ```suggestion
Make sure that you use JDK and JVM versions supporting virtual thread and launch the _dev mode_ with `./mvnw quarkus:dev`:
```

Reviewer: Note that Vale has a point: we might favour the usage of `OpenJDK` over `JDK`

2. When the code invokes the rest client, the virtual thread is blocked, BUT the carrier thread is not blocked (that's the virtual thread _magic touch_).
3. Once the first invocation of the rest client completes, the virtual thread is rescheduled and continues its execution.
4. The second rest client invocation happens, and the virtual thread is blocked again (but not the carrier thread).
5. Finally, when the second invocation of the rest client completes, the virtual thread is rescheduled and continues its execution.
6. The method returns the result. The virtual thread terminates.
7. The result is captured by Quarkus and written in the HTTP response

== Verify pinning using tests

In the `pom.xml,` we added an `argLine` argument to the surefire and failsafe plugins:

[source, xml]
----
<argLine>--enable-preview -Djdk.tracePinnedThreads</argLine>
----

The `-Djdk.tracePinnedThreads` dumps the stack trace if a virtual thread cannot be _unmounted_ smoothly (meaning that it blocks the carrier thread).
That's what we call _pinning_ (more info in xref:./virtual-threads#pinning[the virtual thread reference guide]).

> We recommend enabling this flag in tests.

Reviewer 2: It would be cool if we had some integration with our `@QuarkusTest` that would allow users to retrieve them and assert on them.  Obviously it's unrelated to this PR

Reviewer 2: :)
